lang/src/org/partiql/lang/eval/physical/operators/AggregateOperatorFactory.kt

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2022 Amazon.com, Inc. or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ *  You may not use this file except in compliance with the License.
+ * A copy of the License is located at:
+ *
+ *      http://aws.amazon.com/apache2.0/
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
+ *  language governing permissions and limitations under the License.
+ */
+package org.partiql.lang.eval.physical.operators
+
+import com.amazon.ion.IonValue
+import org.partiql.lang.domains.PartiqlPhysical
+import org.partiql.lang.eval.DEFAULT_COMPARATOR
+import org.partiql.lang.eval.ExprValueType
+import org.partiql.lang.eval.MISSING_ANNOTATION
+import org.partiql.lang.eval.physical.EvaluatorState
+import org.partiql.lang.eval.physical.SetVariableFunc
+import org.partiql.lang.eval.relation.relation
+import org.partiql.lang.util.toIntExact
+import java.util.TreeSet
+
+public abstract class AggregateOperatorFactory(name: String) : RelationalOperatorFactory {
+    public final override val key: RelationalOperatorFactoryKey = RelationalOperatorFactoryKey(RelationalOperatorKind.AGGREGATE, name)
+    public abstract fun create(
+        impl: PartiqlPhysical.Impl,
+        source: RelationExpression,
+        keys: List<CompiledGroupKey>,
+        functions: List<CompiledAggregateFunction>
+    ): RelationExpression
+
+    public abstract fun getFunctionFactory(
+        operator: AggregationKind,
+        name: String
+    ): AggregationFunctionFactory
+}
+
+internal class FullAggregateFactory(name: String) : AggregateOperatorFactory(name) {
+
+    private val aggregationFunctionFactories = mapOf(
+        AggregationFunctionFactoryKey(AggregationKind.MAX, "MAX") to MaxFactory("MAX"),
+        AggregationFunctionFactoryKey(AggregationKind.MIN, "MIN") to MinFactory("MIN"),
+        AggregationFunctionFactoryKey(AggregationKind.COUNT, "COUNT") to CountFactory("COUNT"),
+        AggregationFunctionFactoryKey(AggregationKind.SUM, "SUM") to SumFactory("SUM"),
+        AggregationFunctionFactoryKey(AggregationKind.GROUP_AS, "GROUP_AS") to GroupAsFactory("GROUP_AS"),
+        AggregationFunctionFactoryKey(AggregationKind.AVG, "AVG") to AverageFactory("AVG"),
+    )
```

Comment:
This is interesting. 
In my opinion, this approach is similar to our aggregate function implementation in evaluating compiler. Only that now with the planner, we can have multiple `aggregate` operator, i.e., sort based aggregate, hash based aggregate, and we have to map those function to each one of the operator implementation. 

It will be interesting to design a factory so that we can essentially pass in the factory and called `.createFunction()`. This is actually something that I am experimenting with the window functions. 

Anyways, just want to share some of my thoughts here. I am good with this implementation so far. No blocker.